### PR TITLE
Support multiple patient email addresses - Frontend [3/3]

### LINKED ISSUE #1951: Store multiple emails for a patient
### Description
Currently, users are only able to store one email with the patient. When we introduce emailing results to patients this will be an issue when there are multiple people (i.e. parent/guardian) who will want to or have a right to the test result.

### Proposed solution 
Allow users to add multiple email addresses instead of just one, and let them receive results at all numbers or no numbers. This should be have similarly to multiple phone numbers. 

- [ ] Add multiple emails in self-registration and add patient

- [ ] Update self-registration, add patient, and time of test questionnaire to ask if a patient wants to receive results at all emails or no emails

### PR DESCRIPTION
**Pull request by nathancrtr**
## Related Issue or Background Info

- Closes #1951 

## Changes Proposed
- Adds a `ManageEmails` component for use in patient self-registration and the admin "edit patient" forms
    - Overall this _very_ closely resembles the component to manage multiple phone numbers - this was by design, but there's a decent amount of copy-and-paste. As we go forward we may find a way we like to DRY them out.

## Additional Information
- N/A

### Add/edit person form
https://user-images.githubusercontent.com/27730981/141360991-c1213799-6338-42e2-af9d-aaba6da77e77.mov

### AoE form
Screen Shot 2021-11-11 at 2 53 28 PM

### Re-send test results via email
Screen Shot 2021-11-12 at 6 03 45 PM

### Multiple emails are sent
https://user-images.githubusercontent.com/27730981/141379923-06a4db2f-eddf-4444-9bf9-e0b47649f7c4.mov

## Checklist for Author and Reviewer
- [ ] Getting the yup validation right was a bit of a pain - did I miss any sneaky edge cases?

### UI
- [ ] Any changes to the UI/UX are approved by design 
- [ ] Any new or updated content (e.g. error messages) are approved by design 

### Testing
- [ ] Includes a summary of what a code reviewer should verify

### Changes are Backwards Compatible
- [x] Database changes are submitted as a separate PR
  - [x] Any new tables that do not contain PII are accompanied by a GRANT SELECT to the no-PHI user
  - [x] Any changes to tables that have custom no-PHI views are accompanied by changes to those views
        (including re-granting permission to the no-PHI user if need be)
  - [x] Liquibase rollback has been tested locally using `./gradlew liquibaseRollbackSQL` or `liquibaseRollback`
  - [x] Each new changeset has a corresponding [tag](https://docs.liquibase.com/change-types/community/tag-database.html)
- [x] GraphQL schema changes are backward compatible with older version of the front-end

### Security
- [x] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [x] Any dependencies introduced have been vetted and discussed

## Cloud
- [x] DevOps team has been notified if PR requires ops support
- [x] If there are changes that cannot be tested locally, this has been deployed to our Azure `test`, `dev`, or `pentest` environment for verification

### CODE REVIEW

File: frontend/src/app/patients/AddPatient.tsx [hunk 1]
```diff
@@ -247,6 +250,7 @@ const AddPatient = () => {
             return phoneNumber && phoneNumber.number && phoneNumber.type;
           }
         ),
+        emails: dedupeAndCompactStrings(person.emails || []),
```
Comment (nathancrtr) [inline frontend/src/app/patients/AddPatient.tsx:253]: I'm actually not sure how we want to handle submissions with blank or duplicate emails. This line will remove them - is that desired?

**Comment by zdeveloper** [comment]
It seems like the backend is not returning the emails correctly.

image
image

File: frontend/src/app/constants/constants.d.ts [hunk 2]
```diff
@@ -594,6 +594,7 @@ interface PersonUpdate extends Address {
   county: string;
   country: string;
   email: string;
```
Comment (zdeveloper) [inline frontend/src/app/constants/constants.d.ts:596]: remove this fields, we dont need it in the frontend

File: frontend/src/app/patients/AddPatient.test.tsx [hunk 3]
```diff
@@ -124,6 +124,7 @@ describe("AddPatient", () => {
               ],
               role: null,
               email: null,
```
Comment (zdeveloper) [inline frontend/src/app/patients/AddPatient.test.tsx:126]: remove

**Comment by nathancrtr** [comment]
@zdeveloper I'm not seeing that... at the risk of asking a dumb question, do you need to restart your backend or something?

Screen Shot 2021-11-15 at 2 35 32 PM

File: frontend/src/app/patients/AddPatient.tsx [hunk 4]
```diff
@@ -41,6 +41,7 @@ export const EMPTY_PERSON: Nullable<PersonFormData> = {
   phoneNumbers: null,
   county: null,
   email: null,
```
Comment (zdeveloper) [inline frontend/src/app/patients/AddPatient.tsx:43]: remove email

File: frontend/src/app/patients/AddPatient.tsx [hunk 5]
```diff
@@ -86,6 +87,7 @@ export const ADD_PATIENT = gql`
     $role: String
     $lookupId: String
     $email: String
```
Comment (zdeveloper) [inline frontend/src/app/patients/AddPatient.tsx:89]: remove email

File: frontend/src/app/patients/personSchema.ts [hunk 6]
```diff
@@ -158,6 +159,37 @@ export function isValidBirthdate18n(t: TFunction) {
   };
 }
 
+function isValidEmail18n(t: TFunction) {
+  return function isValidEmail(this: TestContext, email: string | undefined) {
```
Comment (zdeveloper) [inline frontend/src/app/patients/personSchema.ts:163]: return anonymous arrow function

File: frontend/src/app/patients/personSchema.ts [hunk 7]
```diff
@@ -158,6 +159,37 @@ export function isValidBirthdate18n(t: TFunction) {
...
+    }
+
+    return true;
+  };
+}
+
+function areValidEmails18n(t: TFunction) {
+  const validator = isValidEmail18n(t);
+
+  return function areValidEmails(emails: string[] | null | undefined) {
```
Comment (zdeveloper) [inline frontend/src/app/patients/personSchema.ts:185]: return anonymous arrow function

**Review by zdeveloper — CHANGES_REQUESTED** [review]
This design doesn't match the design in the story
image
it should be this
image

we dont have any notion of primary email address, meaning lets only use the emails field in the frontend. that way we can remove the email field from the backend in the future PR

**Comment by nathancrtr** [comment]
@Rebecca-LM now that we support multiple email addresses, this text will need to be changed:

> Would you like to receive results at this email?

Probably the only question there is whether "email address" is more correct than "email" in this context.

**Comment by Rebecca-LM** [comment]
> @Rebecca-LM now that we support multiple email addresses, this text will need to be changed:
> 
> > Would you like to receive results at this email?
> 
> Probably the only question there is whether "email address" is more correct than "email" in this context.

@nathancrtr do you have a screenshot of what it looks like now for context? Without seeing that I'd recommend changing to this. Any singular result is just result. Multiple results are results.

"Would you like your test result sent to this email?"

**Comment by nathancrtr** [comment]
> > @Rebecca-LM now that we support multiple email addresses, this text will need to be changed:
> > > Would you like to receive results at this email?
> > 
> > 
> > Probably the only question there is whether "email address" is more correct than "email" in this context.
> 
> @nathancrtr do you have a screenshot of what it looks like now for context? Without seeing that I'd recommend changing to this. Any singular result is just result. Multiple results are results.
> 
> "Would you like your test result sent to this email?"

Yes! See below.

Screen Shot 2021-11-15 at 2 59 54 PM

I think this is what you're saying, do I have it right:

**One email address** - "Would you like your test result sent to this email?"
**Multiple email addresses** - "Would you like your test result sent to these emails?"

**Comment by nathancrtr** [comment]
@zdeveloper good call on the "primary" and "additional" language. I've removed those words from the labels, although the person form component still maintains a notion of "primacy" for some UI reasons (e.g. so you can't delete every email field from the form). See 6ae9b5e.

**Comment by Rebecca-LM** [comment]
> > > @Rebecca-LM now that we support multiple email addresses, this text will need to be changed:
> > > > Would you like to receive results at this email?
> > > 
> > > 
> > > Probably the only question there is whether "email address" is more correct than "email" in this context.
> > 
> > 
> > @nathancrtr do you have a screenshot of what it looks like now for context? Without seeing that I'd recommend changing to this. Any singular result is just result. Multiple results are results.
> > "Would you like your test result sent to this email?"
> 
> Yes! See below.
> 
> Screen Shot 2021-11-15 at 2 59 54 PM
> 
> I think this is what you're saying, do I have it right:
> 
> **One email address** - "Would you like your test result sent to this email?" **Multiple email addresses** - "Would you like your test result sent to these emails?"

Thanks for the screenshot and yes, that sounds good to me! Thanks @nathancrtr

File: frontend/src/app/constants/constants.d.ts (hunk 2 above)
Comment (nathancrtr) [reply frontend/src/app/constants/constants.d.ts:596]: I reckon we might as well. I originally included this is a step in the ["clean-up" ticket](https://app.zenhub.com/workspaces/simplereport-better-data-team-605b43d51ea9f700119a48ee/issues/cdcgov/prime-simplereport/2974) for this feature, but I believe we can remove it cleanly here.

Comment (nathancrtr) [reply frontend/src/app/constants/constants.d.ts]: Done in f6323fd.

File: frontend/src/app/patients/AddPatient.test.tsx (hunk 3 above)
Comment (nathancrtr) [reply frontend/src/app/patients/AddPatient.test.tsx]: Done in f6323fd.

File: frontend/src/app/patients/AddPatient.tsx (hunk 4 above)
Comment (nathancrtr) [reply frontend/src/app/patients/AddPatient.tsx]: Done in f6323fd.

File: frontend/src/app/patients/AddPatient.tsx (hunk 5 above)
Comment (nathancrtr) [reply frontend/src/app/patients/AddPatient.tsx]: Done in f6323fd.

File: frontend/src/app/patients/personSchema.ts (hunk 6 above)
Comment (nathancrtr) [reply frontend/src/app/patients/personSchema.ts:163]: Is this a preference thing or is there a footgun I don't know? Anonymous functions are great for a lot of things, but I think names are perfectly sensible for returning functions that may be called directly!

```
function named() {
  return function namedFn() {
    throw new Error('I have a name');
  }
}

function anon() {
  return () => {
    throw new Error('I have no name');
  }
}

const n = named();
const a = anon();

try {
  n();
} catch(e) {
  console.log(e);
}

try {
  a();
} catch(e) {
  console.log(e);
}
```

```
➜  prime-simplereport git:(nac/2848-bugix-device-update) ✗ node things.js
Error: I have a name
    at namedFn (/Users/nathancarter/repos/prime-simplereport/things.js:3:11)
    at Object.<anonymous> (/Users/nathancarter/repos/prime-simplereport/things.js:17:3)
Error: I have no name
    at /Users/nathancarter/repos/prime-simplereport/things.js:9:11
    at Object.<anonymous> (/Users/nathancarter/repos/prime-simplereport/things.js:23:3)
```

File: frontend/src/app/patients/AddPatient.tsx (hunk 1 above)
Comment (zdeveloper) [reply frontend/src/app/patients/AddPatient.tsx:250]: yeah that seems acceptable to me

File: frontend/src/app/patients/personSchema.ts (hunk 6 above)
Comment (zdeveloper) [reply frontend/src/app/patients/personSchema.ts:163]: yeah, they're basically equivalent, but with returning a named function, the name is never used, thus it serves no purpose, hence why anonymous functions are a thing.

Comment (zdeveloper) [reply frontend/src/app/patients/personSchema.ts:163]: I am not sure what you mean by `"I think names are perfectly sensible for returning functions that may be called directly!"`,  `isValidEmail` is not accessible directly. you have to use `isValidEmail18n`

**Review by zdeveloper — APPROVED** [review]
Thank you for the great work on this!

**Pull request merged**